webdriver: (W3C) Remove invalid capabilities from alwaysMatch

Varun-garg opened the pull request:
## Proposed changes

In `alwaysMatch` section of W3C capabilities we should only use the capabilities defined [here](https://w3c.github.io/webdriver/#capabilities) and the [extension capabilities](https://w3c.github.io/webdriver/#dfn-extension-capability). This PR will help in running tests on selenium servers where this validation is enforced.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

Signed-off-by: Varun Garg <varun.ga@browserstack.com>

### Reviewers: @webdriverio/technical-committee

christian-bromann commented on packages/webdriver/src/constants.js:
```diff
@@ -86,5 +86,13 @@ export const DEFAULTS = {
+    /**
+     * w3c allowed capabilities
+     */
+    w3cAllowedCaps: [
```
this property should not be part of the option defaults

christian-bromann commented on packages/webdriver/src/index.js:
```diff
@@ -37,6 +37,13 @@ export default class WebDriver {
              */
             : [{ alwaysMatch: params.capabilities, firstMatch: [{}] }, params.capabilities]
 
+        if (w3cCaps.alwaysMatch) {
```
this is always true, just strip them directly

christian-bromann commented on packages/webdriver/src/index.js:
```diff
@@ -37,6 +37,11 @@ export default class WebDriver {
              */
             : [{ alwaysMatch: params.capabilities, firstMatch: [{}] }, params.capabilities]
 
+        w3cCaps.alwaysMatch = JSON.parse(JSON.stringify(w3cCaps.alwaysMatch))
```
There is no reason to do that

Varun-garg commented:
Thanks for reviewing @christian-bromann . Not cloning alwaysMatch would also remove invalid caps from `params.capabilities`.  This would break backward compatibility with the old/extended caps/jars.

christian-bromann commented on packages/webdriver/src/index.js:
```diff
@@ -37,6 +38,14 @@ export default class WebDriver {
              */
             : [{ alwaysMatch: params.capabilities, firstMatch: [{}] }, params.capabilities]
 
+        w3cCaps.alwaysMatch = JSON.parse(JSON.stringify(w3cCaps.alwaysMatch))
```
Can we have it directly in the object creation? like
```js
/**
 * otherwise assume they passed in jsonwp-style caps (flat object)
 */
{ alwaysMatch: Object.keys(params.capabilities)
    .filter(key => !W3C_ALLOWED_CAPABILITIES.includes(key) && !key.match(/^[\w-]+:.*$/))
    .reduce((obj, key) => {
      obj[key] = params.capabilities[key]
      return obj
    }, {}), firstMatch: [{}] },
  params.capabilities]
```
instead of overwriting it afterwards.

christian-bromann commented:
ping @Varun-garg

Varun-garg commented:
Hey @christian-bromann. Sorry I for being late. While your solution looks more elegant, there were 2 issues.

1. Lines were increased after adding warning.
```
: [{
    alwaysMatch: Object.keys(params.capabilities)
        .filter(key => {
            if (W3C_ALLOWED_CAPABILITIES.includes(key) || key.match(/^[\w-]+:.*$/)) {
                return true
            } else {
                log.warn(`Skipping capability ${key} from alwaysMatch.`)
                return false
            }
        })
        .reduce((obj, key) => {
            obj[key] = params.capabilities[key]
            return obj
        }, {}), firstMatch: [{}]
}, params.capabilities]
```

2. We aren't filtering firstMatch if  `params.capabilities.alwaysMatch` exists.

I have rebased and fixed linter errors in test file.

christian-bromann commented:
> 1. Lines were increased after adding warning.

That is no problem, you can shorten this by doing:
```js
: [{
    alwaysMatch: Object.keys(params.capabilities)
        .filter(key => (W3C_ALLOWED_CAPABILITIES.includes(key) || key.match(/^[\w-]+:.*$/))
        .reduce((obj, key) => {
            obj[key] = params.capabilities[key]
            return obj
        }, {}), firstMatch: [{}]
}, params.capabilities]
```

> 1. We aren't filtering firstMatch if  `params.capabilities.alwaysMatch` exists.

If someone explicitly uses W3C conform capabilities and adds an invalid capability into it, we should not strip this invalid capability and let have the test fail.

> I have rebased and fixed linter errors in test file.

Build is still failing but I am not sure if it is because of this patch or some other reason.

Varun-garg commented:
> That is no problem, you can shorten this by doing:
Sorry for sounding stupid, I think you skipped the warnings.

> If someone explicitly uses W3C conform capabilities and adds an invalid capability into it, we should not strip this invalid capability and let have the test fail.
Noted this.

> Build is still failing but I am not sure if it is because of this patch or some other reason.
Will investigate this.

christian-bromann commented:
> Sorry for sounding stupid, I think you skipped the warnings.

Not sure if warnings are needed here

Varun-garg commented:
To let the users know of capabilities that we are dropping.
 
> I think it would be great to log a warning with capabilities that had to be removed because they are not W3C conform (only if there are any).

Varun-garg commented:
I have pushed some changes, I doubt if travis errors are caused by this change.

master:
```
Test Suites: 15 failed, 113 passed, 128 total
Tests:       36 failed, 997 passed, 1033 total
Snapshots:   0 total
```

current branch:
```
Test Suites: 15 failed, 113 passed, 128 total
Tests:       36 failed, 999 passed, 1035 total
Snapshots:   0 total
```

christian-bromann commented on packages/webdriver/src/constants.js:
```diff
@@ -88,3 +88,11 @@ export const DEFAULTS = {
+/**
+ * w3c allowed capabilities
+ */
+export const W3C_ALLOWED_CAPABILITIES = [
```
there are a bunch of caps here that are not allowed capabilities, see here https://w3c.github.io/webdriver/#capabilities

Varun-garg replied:
yup, just fixed 😄

christian-bromann replied:
I still see invalid caps in there

Varun-garg replied:
only to fix travis tests, they were being used in the existing webdriverio code.

Varun-garg commented:
Hey @christian-bromann there seemed to be some problem with my setup. I did need to whitelist some more capabilities not listed [here](https://w3c.github.io/webdriver/#capabilities). We may want to use the [extension capabilities fromat](https://w3c.github.io/webdriver/#dfn-extension-capability) to avoid differences with other drivers.

christian-bromann commented:
> We may want to use the [extension capabilities fromat](https://w3c.github.io/webdriver/#dfn-extension-capability) to avoid differences with other drivers.

Yeah but this needs to be defined on the users end. If the user provides cap extensions it will not be stripped from the caps.

Varun-garg commented:
> > We may want to use the [extension capabilities fromat](https://w3c.github.io/webdriver/#dfn-extension-capability) to avoid differences with other drivers.
> 
> Yeah but this needs to be defined on the users end. If the user provides cap extensions it will not be stripped from the caps.

Yup, I mean we can whitelist the present extensions (like these 4 caps) but ask users in future to use the extended format for new caps

christian-bromann commented:
> Yup, I mean we can whitelist the present extensions (like these 4 caps) but ask users in future to use the extended format for new cap

But this would speak against this whole PR which strips them away. Why should we not strip these away?

Varun-garg commented:
Hey, I also don't want those invalid caps. Remember the travis tests that were failing? I fixed them by finding these capabilities in the test cases and whitelisting them.

christian-bromann commented:
> Hey, I also don't want those invalid caps. Remember the travis tests that were failing? I fixed them by finding these capabilities in the test cases and whitelisting them.

Well you might need to modify some existing tests to match your changes.

christian-bromann commented:
ping @Varun-garg

christian-bromann commented:
@Varun-garg anything I can help with to move this forward?

Varun-garg commented:
@christian-bromann I will have to look into tests of other modules that are failing to remove those invalid caps. Will pick this back shortly

nextlevelbeard commented:
Adding some info to this because it might be relevant to the approach used.

Some invalid capabilities **need** to be passed to 3rd party services in order to fully use them and prefixes don't solve this issue, i.e. specifying
```javascript
capabilities: [
  {
    browserName: 'safari',
    'browserstack.local': true
  }
]
```
is needed to run tests in Safari on BrowserStack when the application under test is on the localhost or can only be accessed by the localhost.

maggiesavovska commented:
Hi! For crossbrowsertesting.com I pass "acceptSslCerts" cap. Can we add that as a valid cap? Thanks!

christian-bromann commented:
> Can we add that as a valid cap?

It is a valid cap

christian-bromann commented:
Closing due to inactivity. Please create a new PR if you think this patch is valuable to project. Thanks!
